### PR TITLE
PERF: cythonize kendall correlation

### LINKED ISSUE #28329: DataFrame.corr(method="kendall") calculation is slow
```python
import numpy as np
import pandas as pd

df = pd.DataFrame(np.random.randn(1000, 300))

df.corr(method="kendall")
# 21.6 s ± 686 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

`DataFrame.corr(method="kendall")` doesn't scale particularly well, perhaps because it's the only named correlation method that isn't Cythonized at the moment (we just call `kendalltau` from `scipy` repeatedly in a Python for loop: https://github.com/pandas-dev/pandas/blob/master/pandas/core/frame.py#L7454).  It may be worthwhile to try to implement something more efficient within `_libs/algos.pyx`.

Relevant discussion: https://github.com/pandas-dev/pandas/pull/28151

### PR DESCRIPTION
**Pull request by lithomas1**
- [x] closes #28329 
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

ASV's. 

       before           after         ratio
     [a6bc6ecc]       [1eb86441]
     <perf-cythonize-kendall~2^2>       <perf-cythonize-kendall>
-             81M            57.6M     0.71  stat_ops.Correlation.peakmem_corr_wide('kendall')
-      1.41±0.02s          362±7ms     0.26  stat_ops.Correlation.time_corr_wide_nans('kendall')
-      1.57±0.04s          398±4ms     0.25  stat_ops.Correlation.time_corr_wide('kendall')
-        34.6±4ms      8.75±0.06ms     0.25  stat_ops.Correlation.time_corr('kendall')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.

### CODE REVIEW

File: pandas/_libs/algos.pyx [hunk 1]
```diff
@@ -393,6 +393,77 @@ def nancorr_spearman(ndarray[float64_t, ndim=2] mat, Py_ssize_t minp=1) -> ndarr
...
+                result[xi, yi] = NaN
+                result[yi, xi] = NaN
+            else:
+                # Get columns and order second column using 1st column ranks
+                if not valid.all():
+                    valid_cols = ranked_mat[valid.nonzero()][:, [xi, yi]]
+                else:
+                    valid_cols = ranked_mat[:, [xi, yi]]
+                # Unfortunately we have to sort here, since we can have tied indices
+                col = valid_cols[:, 1][valid_cols[:, 0].argsort()]
```
Comment (jreback) [inline pandas/_libs/algos.pyx:444]: this is going to make your *also* super slow

**Comment by lithomas1** [comment]
@jreback this is now ready for review.
cc @jbrockmendel and @dsaxton(if you guys are interested).

**Review by jreback — APPROVED** [review]
lgtm cc @jbrockmendel if any comments

File: pandas/_libs/algos.pyx [hunk 2]
```diff
@@ -393,6 +393,81 @@ def nancorr_spearman(ndarray[float64_t, ndim=2] mat, Py_ssize_t minp=1) -> ndarr
     return result
 
 
+# ----------------------------------------------------------------------
+# Kendall correlation
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def nancorr_kendall(ndarray[float64_t, ndim=2] mat, Py_ssize_t minp=1) -> ndarray:
```
Comment (jreback) [inline pandas/_libs/algos.pyx:402]: actually, can you add a small doc-string with Parameters and Returns (esp shapes / dtypes) & put a link the kendall algo (wikipedia fine)

**Comment by lithomas1** [comment]
A small slightly unrelated change that I made in this PR was in the docstring of the min_periods arg for corr. It used to state that min_periods was only valid for spearman and pearson correlations, however, I believe that it always worked for kendall correlation.

File: pandas/_libs/algos.pyx [hunk 3]
```diff
@@ -393,6 +393,99 @@ def nancorr_spearman(ndarray[float64_t, ndim=2] mat, Py_ssize_t minp=1) -> ndarr
     return result
 
 
+# ----------------------------------------------------------------------
+# Kendall correlation
+# Wikipedia article: https://en.wikipedia.org/wiki/Kendall_rank_correlation_coefficient
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def nancorr_kendall(ndarray[float64_t, ndim=2] mat, Py_ssize_t minp=1) -> ndarray:
```
Comment (jbrockmendel) [inline pandas/_libs/algos.pyx:402]: we usually dont mix-and-match cython-vs-python style annotations.  i know this is tough bc the python-style won't allow `ndarray[float64_t, ndim=2]`.  not sure if there's a better alternative

File: pandas/_libs/algos.pyx [hunk 4]
```diff
@@ -393,6 +393,99 @@ def nancorr_spearman(ndarray[float64_t, ndim=2] mat, Py_ssize_t minp=1) -> ndarr
...
+    """
+    Perform kendall correlation on a 2d array
+
+    Parameters
+    ----------
+    mat : Array to compute kendall correlation on
```
Comment (jbrockmendel) [inline pandas/_libs/algos.pyx:408]: `mat : np.ndarray[float64_t, ndim=2]`

File: pandas/_libs/algos.pyx [hunk 5]
```diff
@@ -393,6 +393,99 @@ def nancorr_spearman(ndarray[float64_t, ndim=2] mat, Py_ssize_t minp=1) -> ndarr
...
+    minp : int, default 1
+        Minimum number of observations required per pair of columns
+        to have a valid result.
+
+    Returns
+    -------
+    numpy.ndarray
```
Comment (jbrockmendel) [inline pandas/_libs/algos.pyx:415]: dtype?

File: pandas/_libs/algos.pyx [hunk 6]
```diff
@@ -393,6 +393,99 @@ def nancorr_spearman(ndarray[float64_t, ndim=2] mat, Py_ssize_t minp=1) -> ndarr
...
+                total_discordant = 0
+                for j in range(n_obs - 1):
+                    currj = col[j]
+                    # Count num concordant and discordant pairs
+                    n_concordant = 0
+                    for k in range(j, n_obs):
+                        if col[k] > currj:
+                            n_concordant += 1
+                    total_concordant += n_concordant
+                    total_discordant += (n_obs-1-j-n_concordant)
```
Comment (jbrockmendel) [inline pandas/_libs/algos.pyx:472]: whitespace around operators

**Comment by jbrockmendel** [comment]
couple nitpicks, generally looks good.  i take it we have sufficient testing?

File: pandas/_libs/algos.pyx (hunk 3 above)
Comment (lithomas1) [reply pandas/_libs/algos.pyx:402]: Right now, the function signature for this function is the same as the one for nancorr_spearman. However, I notice that other functions have left out the `` -> ndarray`` part(including nancorr_pearson). Should I remove that from this function?

**Comment by lithomas1** [comment]
@jbrockmendel I think testing coverage should be fine. There is a test that compares the results between scipy and our implementation. (see https://github.com/pandas-dev/pandas/blob/master/pandas/tests/frame/methods/test_cov_corr.py#L89)

File: pandas/_libs/algos.pyx (hunk 3 above)
Comment (jreback) [reply pandas/_libs/algos.pyx:402]: right but we are currently not checking the .pyx right so this is moot for now?

Comment (jbrockmendel) [reply pandas/_libs/algos.pyx:402]: yah its not a big deal

**Comment by jreback** [comment]
thanks @lithomas1 very nice!

**Pull request merged**